Implement deterministic wallet

kanzure opened the pull request:
Here's an implementation of a deterministic wallet.

kanzure committed:
better log messages into the wallet

kanzure committed:
implement deterministic wallet

maaku commented:
utACK

However I think I would redo the way secrets are calculated. Make it something like the following:

```
tag = hashlib.sha256("webcashwallet").digest()
new_secret = hashlib.sha256(tag + tag)
new_secret.update(master_secret) # <- binary, not hex!
new_secret.update(struct.pack(">Q", chain_code)) # big-endian
new_secret.update(struct.pack(">Q", walletdepth))
return new_secret.hexdigest()
```

So the secret derivation function is a tagged sha256 hash (which can be made efficient with midstate resumption) that is 384 bits long, which is short enough to fit in a single sha256 round, padding included. Also the preimage format is fixed in size and binary encoded, which makes possible future zero-knowledge proofs easier to construct.

The `chain_code` is just a 64 bit number that can be used to differentiate different secrets. Use a different chain code for payments, for example, so the wallet can recover them but automatically mark them as separate. Maybe use a different chain code for secrets used to receive payments or mining income too, vs. change? I'm less certain about this.

kanzure committed:
implement chaincodes and recovery for different chaincodes

kanzure committed:
don't recover payments into the wallet

kanzure committed:
harden the healthcheck command

kanzure committed:
also use the deterministic wallet in the default mining client

kanzure committed:
introduce separate chaincode for mining

maaku commented:
Seems to have broken wallet setup?

```
$ python3 walletclient.py setup  
Didn't find an existing webcash wallet, making a new one called default_wallet.webcash
Generating a new wallet with a new master secret...
Traceback (most recent call last):
  File "/<snip>/walletclient.py", line 161, in <module>
    ask_user_for_legal_agreements(webcash_wallet)
NameError: name 'ask_user_for_legal_agreements' is not defined
```

kanzure committed:
create initial webcash wallet in __main__

kanzure commented:
fixed by moving that chunk of code into `__main__`

maaku commented on miner.py:
```diff
@@ -54,8 +54,8 @@ def mine():
-    keep = secrets.token_hex(32)
-    subsidy = secrets.token_hex(32)
+    keep = generate_new_secret(webcash_wallet, chain_code="MINING")
+    subsidy = generate_new_secret(webcash_wallet, chain_code="PAY")
```
Note: every time you run the miner it is going to increment the MINING and PAY wallet depths. Do this 20 times without finding a PoW solution and the wallet would be unrecoverable. Now mining already locks the wallet, so perhaps these should "peek" at the next secret values and then only increment the wallet debts if a solution is found?

On the other hand, that would make mining deterministic, such that if you re-start the miner it will start looking at the same work again. We really ought to put a timestamp in the mining payload structure.

maaku replied:
This isn't resolved, but fixing it might be a bit more complicated than is warranted for this PR. I'm marking my review as 'approved' in case you want to handle this separately.

kanzure replied:
How about this? 7a62404c178d3335c7bc1a973b9c955a54994a82

maaku replied:
Looks good, although I didn't test it.

kanzure replied:
> On the other hand, that would make mining deterministic, such that if you re-start the miner it will start looking at the same work again. We really ought to put a timestamp in the mining payload structure.

67737d33bdd57ca0c19a2e8ef25585fb940b6a8e

maaku replied:
I’ll update webminer to do the same thing.

maaku commented on miner.py:
```diff
@@ -130,7 +128,7 @@ def mine():
 
             print(f"I have created {mining_amount} webcash. Securing secret.")
-            new_webcash = SecretWebcash(amount=mining_amount_remaining + previous_amount, secret_value=secrets.token_hex(32))
+            new_webcash = SecretWebcash(amount=mining_amount_remaining + previous_amount, secret_value=generate_new_secret(webcash_wallet, chain_code="MINING"))
```
This should be CHANGE, I think.

kanzure replied:
why CHANGE? isn't this the mined amount?

maaku replied:
Well the idea is that if you add up all the 'MINING' outputs ever used, you get the total mined by the wallet. Likewise if you add up the 'RECEIVE' you get the total input, 'PAY' for total payments. We already have a mining input from the mining payload submitted above. This is just re-claiming the output into the wallet so the mining secret isn't kept long-term, which is a wallet-internal (aka 'CHANGE') operation.

Even if you ignore the accounting argument, this preserves an important semantic property: any 'MINING' output comes from a mining report and therefore ought to be claimed right away. In principle, if 'recover' found any 'MINING' chain code outputs, it ought to do the same thing. The 'MINING' chain code is only used for mining reports.

maaku commented on walletclient.py:
```diff
@@ -41,11 +44,85 @@
+
+        save_webcash_wallet(webcash_wallet)
+    else:
+        new_secret = secrets.token_hex(length)
```
AFAICT this is only used to generate the new master secret, right? Maybe this should be moved to `generate_new_master_secret`

maaku commented on walletclient.py:
```diff
@@ -151,11 +245,115 @@ def check():
+    # gaplimit is the maximum window span that will be used, on the assumption
+    # that any valid webcash will be found within the last item plus gaplimit
+    # number more of the secrets.
+    gaplimit = 20
```
Maybe should be a configurable option?

maaku commented on walletclient.py:
```diff
@@ -151,11 +245,115 @@ def check():
+            check_webcashes = {}
+            for x in range(current_walletdepth, current_walletdepth + gaplimit):
+                secret = generate_new_secret(webcash_wallet, chain_code=chain_code, walletdepth=x)
+                webcash = SecretWebcash.deserialize("e1:secret:" + secret)
```
nit: why is this "e1:..." ? I thought 0 was the stand-in for "unknown amount"

kanzure replied:
server returns an error for 0 because 0 is an invalid amount, this could be fixed if it's a problem tho

maaku replied:
Not a problem now. But maybe something to change in the future.

maaku reviewed (changes requested):
Nothing serious came up on the review. Ready to merge I think after the things I noted are considered.

maaku commented:
Ok I also tested wallet recovery. One thing that tripped me up right away was there is no way to recover from an explicit written down master seed value. I don't think this should hold us up because there's no way to extract a master seed value anyway except by doing wallet surgery in the first place. But my mental model was originally that there would be some sort of "recovery code" I could input to the `walletclient.py recover` API to rebuild the wallet, but rather `recover` expects that it is given an older version of the wallet, perhaps an offline backup. Maybe a future API will allow outputting a human-readable recovery seed and recovering from that.

Anyway, I put some funds into a new wallet, then further broke that output into smaller change via `pay` requests, only some of which I claimed back into the wallet. I then created a new wallet and copy-pasted the old master secret into the appropriate point in the wallet file and did a recovery. It picked up all the received funds and wallet-internal outputs, and notified me of the outstanding payments that hadn't been claimed. Afterwards the wallets were an identical match (checked by pretty-printing the json and then diff'ing the results), except for the log which was missing from the recovered wallet (expected) and the exact order of the webcash.

kanzure commented:
One thing to note is that the way I'm using `gaplimit` is technically incorrect... not sure if that's a big deal. It's potentially checking up to `(gaplimit * 2) - 1` more items instead of the `gaplimit` amount.

maaku commented:
Oh, and I attempted recovery on a fully used wallet, and found that the walletdepths were correctly recovered. This is good and essential so we don't reuse payment codes!

maaku commented:
I don't think that's incorrect per se, just a unclear variable name. It's not an upper limit, but rather a lower limit. You can guarantee handling gaps of up to 19 for `gaplimit=20` (I think? might be off-by-1), but it could be up to 39.

kanzure committed:
use the CHANGE chaincode for secret derivation during mining replacement

https://github.com/kanzure/webcash/pull/7#discussion_r815482198

kanzure committed:
refactor generate_new_master_secret to simplify secret generation

kanzure committed:
add gaplimit as a command line option for recover

https://github.com/kanzure/webcash/pull/7#discussion_r815478935

kanzure committed:
miner should only save new secrets once mined

https://github.com/kanzure/webcash/pull/7#discussion_r815476400

kanzure committed:
oops, get new secrets after successful mining

kanzure committed:
simplify peek-ahead deterministic secret generation

kanzure committed:
add timestamp into mining payload

With the deterministic derivation of wallet secrets, it's possible that
the mining client might repeat the same work if it was stopped and then
re-started. To fix this, an untrusted timestamp is added.

https://github.com/kanzure/webcash/pull/7#discussion_r815476400

kanzure committed:
fix check vs recover bug in the recover command

kanzure committed:
update check command to reset amounts when they are wrong

kanzure committed:
update check command to remove duplicate webcash

kanzure committed:
small cleanup in miner.py

kanzure commented:
Two things on my mind about this branch:

(1) Should the wallet really have a PAY chaincode? This could either help someone recover funds that were sent as payment that need to be re-sent, and also equally helpful for fraud and maybe it would be better to not include this. So which way to go?

(2) Mining client changes may be incomplete, still thinking about it.
